### PR TITLE
feat: impl ID Allocator 

### LINKED ISSUE #16: feat: impl ID Allocator
**Description**

impl ID Allocator with etcdKV

**Proposal**

can refer to https://github.com/tikv/pd/blob/0c3303e765f2c3580b3ec80ea89982e20cfe5219/server/id/id.go#L31
need to use KV interface, avoid to use etcd client directly

**Additional context**

### PR DESCRIPTION
**Pull request by baojinri**
# Which issue does this PR close?

Closes #16 

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

# How does this change test

### CODE REVIEW

File: server/id/id_impl.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	"path"
+	"sync"
+
+	"github.com/CeresDB/ceresmeta/pkg/log"
+	"github.com/CeresDB/ceresmeta/server/etcdutil"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+)
+
+const allocStep = uint64(1000)
```
Comment (chunshao90) [inline server/id/id_impl.go:17]: defaultAllocStep

File: server/id/id_impl.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+type AllocatorImpl struct {
+	mu       sync.Mutex
+	base     uint64
+	end      uint64
+	client   *clientv3.Client
```
Comment (chunshao90) [inline server/id/id_impl.go:23]: use storage kv interface, remove etcd dep

File: server/id/id_impl.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+type AllocatorImpl struct {
+	mu   sync.Mutex
+	base uint64
+	end  uint64
+	kv   storage.KV
+}
+
+func NewAllocatorImpl(client *clientv3.Client, rootPath string) *AllocatorImpl {
```
Comment (chunshao90) [inline server/id/id_impl.go:29]: func NewAllocatorImpl(kv KV, rootPath string) *AllocatorImpl

File: server/id/id_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	err = allo.Rebase(ctx)
+	re.NoError(err)
+	for i := 0; i < 2010; i++ {
+		value, err := allo.Alloc(ctx)
+		re.NoError(err)
+		re.Equal(uint64(i+3001), value)
+	}
+}
+
+func newTestSingleConfig(t *testing.T) *embed.Config {
```
Comment (chunshao90) [inline server/id/id_test.go:53]: use etcdutil.NewTestSingleConfig

File: server/id/id.go [hunk 5]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package id
+
+import "context"
+
+// Allocator defines the id allocator on the ceresdb cluster meta info.
+type Allocator interface {
+	Alloc(ctx context.Context) (uint64, error)
+	Rebase(ctx context.Context) error
```
Comment (chunshao90) [inline server/id/id.go:10]: add some description

File: server/id/id_impl.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	"go.uber.org/zap"
+)
+
+const (
+	defaultallocStep = uint64(1000)
+	defaultkey       = "alloc_id"
+)
+
+type AllocatorImpl struct {
+	mu   sync.Mutex
```
Comment (chunshao90) [inline server/id/id_impl.go:23]: sync.Mutex

File: server/id/id_impl.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	end  uint64
+	kv   storage.KV
+}
+
+func NewAllocatorImpl(client *clientv3.Client, rootPath string) *AllocatorImpl {
+	return &AllocatorImpl{kv: storage.NewEtcdKV(client, rootPath)}
+}
+
+func (alloc *AllocatorImpl) Alloc(ctx context.Context) (uint64, error) {
+	alloc.mu.Lock()
```
Comment (chunshao90) [inline server/id/id_impl.go:34]: alloc.Lock()

File: server/id/id_impl.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+func (alloc *AllocatorImpl) Rebase(ctx context.Context) error {
+	alloc.Lock()
+	defer alloc.Unlock()
+	return alloc.rebaseLocked(ctx)
+}
+
+func (alloc *AllocatorImpl) rebaseLocked(ctx context.Context) error {
+	key := path.Join(alloc.rootPath, "alloc_id")
```
Comment (chunshao90) [inline server/id/id_impl.go:51]: const

File: server/id/id_test.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+		value, err := allo.Alloc(ctx)
+		re.NoError(err)
+		re.Equal(uint64(i+1), value)
+	}
+	err = allo.Rebase(ctx)
+	re.NoError(err)
+	for i := 0; i < 2010; i++ {
+		value, err := allo.Alloc(ctx)
+		re.NoError(err)
+		re.Equal(uint64(i+3001), value)
```
Comment (chunshao90) [inline server/id/id_test.go:51]: 3001 ？

File: server/id/id_impl.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+	"strconv"
+	"sync"
+
+	"github.com/CeresDB/ceresmeta/pkg/log"
+	"github.com/CeresDB/ceresmeta/server/storage"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+const defaultallocStep = uint64(1000)
```
Comment (chunshao90) [inline server/id/id_impl.go:18]: defaultAllocStep

File: server/id/id_impl.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+		end = 0
+	} else {
+		value, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			return errors.Wrapf(err, "convert string to int err")
+		}
+		end = uint64(value)
+	}
+	end += defaultallocStep
+	err = alloc.kv.Put(ctx, key, fmt.Sprintf("%020d", end))
```
Comment (chunshao90) [inline server/id/id_impl.go:67]: Use txn to avoid data race

File: server/id/id_impl.go [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/CeresDB/ceresmeta/pkg/log"
+	"github.com/CeresDB/ceresmeta/server/storage"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+const defaultallocStep = uint64(1000)
+
+type AllocatorImpl struct {
+	sync.Mutex
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:21]: Add comments to tell which fields this mutex protect.

File: server/id/id_impl.go (hunk 8 above)
Comment (ShiKaiWi) [reply server/id/id_impl.go:51]: This path is actually determinate when `AllocatorImpl` is instantiated so how about keeping the path as a field of `AllocatorImpl`.

File: server/id/id_impl.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+	alloc.Lock()
+	defer alloc.Unlock()
+	return alloc.rebaseLocked(ctx)
+}
+
+func (alloc *AllocatorImpl) rebaseLocked(ctx context.Context) error {
+	key := path.Join(alloc.rootPath, "alloc_id")
+	value, err := alloc.kv.Get(ctx, key)
+	if err != nil {
+		return errors.Wrapf(err, "get base id err, key: %v", key)
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:54]: `get base id failed, key:%v` sounds better and error messages can be refactored in this way:
- err -> failed
- no space between field and field value

File: server/id/id_impl.go [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+	}
+	var end uint64
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:56]:
```go
end := 0
if value != "" {
...
}
```

File: server/id/id_impl.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+	if value == "" {
+		end = 0
+	} else {
+		value, err := strconv.ParseInt(value, 10, 64)
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:60]: `ParseUint` seems a better choice.

File: server/id/id.go [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+package id
+
+import "context"
+
+// Allocator defines the id allocator on the ceresdb cluster meta info.
+type Allocator interface {
+	// Alloc allocs a unique id.
+	Alloc(ctx context.Context) (uint64, error)
+	// Rebase resets the base for the allocator from the persistent window boundary.
+	Rebase(ctx context.Context) error
```
Comment (ShiKaiWi) [inline server/id/id.go:12]: Shall we expose the `Rebase` method to callers? Any cases where callers have to use this method?

File: server/id/id_impl.go [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+	"github.com/pkg/errors"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+)
+
+const defaultAllocStep = uint64(1000)
+
+type AllocatorImpl struct {
+	// lock var base&end when allocator ID
+	sync.Mutex
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:23]: Maybe there is no need to expose this lock.

File: server/id/id_test.go [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/CeresDB/ceresmeta/server/storage"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/server/v3/embed"
+)
+
+const defaultRequestTimeout = time.Second * 10
+
+func TestAlloc(t *testing.T) {
```
Comment (ShiKaiWi) [inline server/id/id_test.go:22]: Need a concurrent test case or at least a TODO for that.

File: server/id/id_impl.go [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+
+func (alloc *AllocatorImpl) idAllocBase(ctx context.Context, value uint64) error {
+	end := value + defaultAllocStep
+	key := path.Join(alloc.rootPath, alloc.key)
+
+	var cmp clientv3.Cmp
+	if value == 0 {
+		cmp = clientv3.Compare(clientv3.CreateRevision(key), "=", 0)
+	} else {
+		cmp = clientv3.Compare(clientv3.Value(key), "=", fmt.Sprintf("%020d", value))
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:79]: - Is it necessary to encode the value by "%020d"? Can we encode it by "%d"?
- Maybe we should add methods called `encodeID`/`decodeID`.

File: server/id/id_impl.go [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+	return alloc.base, nil
+}
+
+func (alloc *AllocatorImpl) rebaseLocked(ctx context.Context) error {
+	value, err := alloc.kv.Get(ctx, alloc.key)
+	if err != nil {
+		return errors.Wrapf(err, "get base id failed, key:%v", alloc.key)
+	}
+	if value != "" {
+		value, err := strconv.ParseUint(value, 10, 64)
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:57]: It is better to define a method called `decodeID` which corresponds to the `encodeID`.

File: server/id/id_impl.go [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+func NewAllocatorImpl(kv storage.KV, rootPath string, key string) *AllocatorImpl {
+	return &AllocatorImpl{kv: kv, key: key, rootPath: rootPath}
+}
+
+func (alloc *AllocatorImpl) Alloc(ctx context.Context) (uint64, error) {
+	alloc.Lock()
+	defer alloc.Unlock()
+
+	if alloc.base == alloc.end {
+		if err := alloc.fastRebaseLocked(ctx); err != nil {
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:40]: Maybe we should print the error msg as a warning.

File: server/id/id_impl.go [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+			}
+		}
+	}
+
+	alloc.base++
+	return alloc.base, nil
+}
+
+func (alloc *AllocatorImpl) rebaseLocked(ctx context.Context) error {
+	value, err := alloc.kv.Get(ctx, alloc.key)
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:52]: Maybe `value` should have a better name, such as `currEnd`.

File: server/id/id_impl.go [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+	end := value + defaultAllocStep
+	key := path.Join(alloc.rootPath, alloc.key)
+
+	var cmp clientv3.Cmp
+	if value == 0 {
+		cmp = clientv3.Compare(clientv3.CreateRevision(key), "=", 0)
+	} else {
+		cmp = clientv3.Compare(clientv3.Value(key), "=", fmt.Sprintf("%020d", value))
+	}
+	opPutBaseID := clientv3.OpPut(key, fmt.Sprintf("%020d", end))
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:81]: Shall we call this `opPutEndID`?

File: server/id/id_impl.go [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	return alloc.idAllocBase(ctx, 0)
+}
+
+func (alloc *AllocatorImpl) fastRebaseLocked(ctx context.Context) error {
+	return alloc.idAllocBase(ctx, alloc.end)
+}
+
+func (alloc *AllocatorImpl) idAllocBase(ctx context.Context, value uint64) error {
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:71]: - Maybe we should call this method `doRebase`?
- Shall we call `value` as `currEnd`?

File: server/id/id_impl.go [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+	end := value + defaultAllocStep
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:72]: `end` should be called `newEnd` if `value` is called `currEnd`.

File: server/id/id_test.go (hunk 18 above)
Comment (baojinri) [reply server/id/id_test.go:22]: TODO

**Review by ShiKaiWi — APPROVED** [review]
LGTM

**Pull request merged**